lib/utils.h

```diff
@@ -269,4 +270,74 @@ std::size_t getArrayLength(const T (& /*unused*/)[size])
     return size;
 }
 
+/** this is meant as a replacement for when we cannot print a pointer via the stream insertion operator (operator<<) for performance reasons */
+// TODO: should give portable value / only used by Tokenizer::dump() and underlying functions - something deterministic would also help with comparing the output
+static inline std::string ptr_to_string(const void* p)
+{
+#if (defined(__APPLE__) && defined(__MACH__))
+    if (p == nullptr)
+        return "0x0";
+#elif !defined(_WIN32) || defined(__MINGW32__)
+    if (p == nullptr)
+        return "0";
+#endif
+
+    static constexpr int ptr_size = sizeof(void*);
+
+#if defined(_WIN32) && !defined(__MINGW32__)
+    // two characters of each byte / contains terminating \0
+    static constexpr int buf_size = (ptr_size * 2) + 1;
+#else
+    // two characters of each byte / contains 0x prefix and contains terminating \0
+    static constexpr int buf_size = (ptr_size * 2) + 2 + 1;
+#endif
+    char buf[buf_size];
+
+    // needs to be signed so we don't underflow in padding loop
+    int idx = sizeof(buf) - 1;
+    buf[idx--] = '\0'; // terminate string
+
+    uintptr_t l = reinterpret_cast<uintptr_t>(p);
+    while (l != 0)
+    {
+        char c;
+        const uintptr_t temp = l % 16; // get the remainder
+        if (temp < 10) {
+            // 0-9
+            c = 48 + temp;
+        }
+        else {
+            // a-f / A-F
+            c = 55 + temp;
+#if !defined(_WIN32) || defined(__MINGW32__)
+            c += 32; // add 32 for lowercase
```

Comment:
there is no technical reason to use lower case as far as I know.

Reply:
It is necessary to match the result of `std::ostringstream`. As mentioned before I will do the portable approach in another PR.

Reply:
if we really want to lowercase, I think that either `c = 'a' + temp;` or `c += 'a' - 'A';` would be more elegant.

Reply:
I don't see why it's important to match std::ostringstream though. the only possible usage I can think of for this function is when generating the dump file or debug output.

Reply:
People might depend on the format. Also I want that change separate so it can easily be reverted if necessary. I will also put it behind a define for a release or two.

Reply:
> if we really want to lowercase, I think that either `c = 'a' + temp;` or `c += 'a' - 'A';` would be more elegant.

I think the `- 10` is better as it highlights that we are dealing with a "base 16"/hex value as input.